### PR TITLE
feat: [IDP-554]: URL Allowlist for backstage

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48680)

### CODE REVIEW

**Comment by vigneswara-propelo** [comment]
trigger commonchecks

**Comment by vigneswara-propelo** [comment]
trigger uts

File: idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    JsonNode rootNode = asJsonNode(yamlString);
+    getReadingNode(rootNode).put(ALLOW_PROPERTY, allowListNode);
+    String config = asYaml(rootNode.toString());
+    createOrUpdateAllowListAppConfig(config, harnessAccount);
+    return hostInfoList;
+  }
+
+  private ObjectNode getReadingNode(JsonNode rootNode) {
+    JsonNode backendNode = rootNode.get(BACKEND_PROPERTY);
+    return (ObjectNode) backendNode.get(READING_PROPERTY);
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java]: should we check if backendNode is not null?

File: idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  private void createOrUpdateAllowListAppConfig(String config, String accountIdentifier) throws Exception {
+    String schema = CommonUtils.readFileFromClassPath(ALLOW_LIST_JSON_SCHEMA_FILE);
+    if (!isValidSchema(config, schema)) {
+      log.error(String.format(INVALID_SCHEMA_FOR_ALLOW_LIST, accountIdentifier));
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java]: should we return from here if the validation fails?

File: idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java (hunk 1 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java:70]: we wont need this check as these nodes are defined in the config file by default

File: idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java (hunk 2 above)
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/allowlist/services/AllowListServiceImpl.java]: done

File: idp-service/src/main/java/io/harness/idp/common/YamlUtils.java [hunk 3]
```diff
@@ -25,4 +26,12 @@ public static String writeObjectAsYaml(Object obj) {
       throw new UnexpectedException("Error writing object as yaml");
     }
   }
+
+  public static <T> T read(String value, Class<T> cls) {
+    try {
+      return YamlPipelineUtils.read(value, cls);
+    } catch (IOException e) {
+      throw new UnexpectedException("Error reading the content");
```
[the remark this answers is not in the document]
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/common/YamlUtils.java]: done

File: idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigManagerServiceImpl.java [hunk 4]
```diff
@@ -83,9 +83,9 @@ public Map<String, Boolean> getAllPluginIdsMap(String accountIdentifier) {
   }
 
   @Override
-  public AppConfig getPluginConfig(String accountIdentifier, String pluginId) {
-    Optional<AppConfigEntity> pluginConfig = appConfigRepository.findByAccountIdentifierAndConfigIdAndConfigType(
-        accountIdentifier, pluginId, ConfigType.PLUGIN);
+  public AppConfig getAppConfig(String accountIdentifier, String configId, ConfigType configType) {
+    Optional<AppConfigEntity> pluginConfig =
```
[the remark this answers is not in the document]
Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigManagerServiceImpl.java]: done

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vikyathharekal** [comment]
Are we planning to add tests for the new APIs and Services?

**Review by vikyathharekal — APPROVED** [review]

**Comment by vigneswara-propelo** [comment]
> Are we planning to add tests for the new APIs and Services?

yes, I will create separate PR for Unit tests

**Review by devesh-harness — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger review

**Review by mankrit-singh — APPROVED** [review]

**Pull request merged**